components/data_stores/actions/add-update-multiple-records/add-update-multiple-records.mjs

```diff
@@ -19,19 +19,76 @@ export default {
     data: {
       label: "Data",
       type: "object",
-      description: "Enter data you'd like to add as an object containing key-value pairs, or reference existing keys and update the values for those existing records. Refer to your existing keys [here](https://pipedream.com/data-stores/).",
+      description: "Enter data you'd like to add as key-value pairs, or reference an object from a previous step using a custom expression (e.g., `{{steps.data.$return_value}}`).",
+    },
+  },
+  methods: {
+    _sanitizeJson(str) {
+      return str.replace(/(['"])?([a-z0-9A-Z_]+)(['"])?:/g, "\"$2\": ");
+    },
+    _parseDataRecursive(obj) {
+      let result = {};
+      if (Array.isArray(obj)) {
+        obj.forEach((o) => {
+          result = {
+            ...result,
+            ...this._parseDataRecursive(o),
+          };
+        });
+      } else if (typeof obj === "object") {
+        result = {
+          ...result,
+          ...obj,
+        };
+      } //else pass, because it is neither object nor array
+      return result;
+    },
+    _parseData(obj) {
+      if (typeof obj === "string") {
+        const sanitizedValue = this._sanitizeJson(obj);
+        const json = JSON.parse(sanitizedValue);
+        return this._parseDataRecursive(json);
+      } else if (typeof obj === "object") {
+        const newObj = {};
+        for (const [
+          key,
+          value,
+        ] of Object.entries(obj)) {
+          try {
+            const sanitizedValue = this._sanitizeJson(value);
+            const json = JSON.parse(sanitizedValue);
+            newObj[key] = this._parseDataRecursive(json);
+          } catch (err) {
+            if (typeof value === "object") {
+              newObj[key] = this._parseDataRecursive(value);
+            } else {
+              newObj[key] = value;
+            }
+          }
+        }
+        return newObj;
+      }
+      throw new Error("Unsupported expression");
     },
   },
   async run({ $ }) {
+    let data;
+    if (typeof this.data !== "object") {
+      try {
+        const sanitizedValue = this._sanitizeJson(this.data);
+        data = JSON.parse(sanitizedValue);
+      } catch (err) {
+        throw new Error("please provide an object or object array!");
+      }
+    }
+    data = this._parseData(this.data);
     for (const [
       key,
       value,
-    ] of Object.entries(this.data)) {
+    ] of Object.entries(data)) {
```

Comment:
I think we can simplify the conversion to this code:

```
const arrayToMap = (arr, map = {}) => {
    if (!Array.isArray(arr)) {
        return typeof(arr) === "object"
            ? arr
            : {}
    }
    for (let i = 0; i < arr.length; i++) {
        if (Array.isArray(arr[i])) {
            arrayToMap(arr[i], map)
        } else if (typeof arr[i] === "object") {
            const keys = Object.keys(arr[i])
            for (let j = 0; j < keys.length; j++) {
                map[keys[j]] = arr[i][keys[j]]
            }
        }  
    }
    
    return map
}
```

You can just call this function passing any value, it will return a hashmap with the keys.

Reply:
Please try this

```suggestion
    const arrayToMap = (data) => {
  if (!Array.isArray(data)) {
      return typeof(data) === "object"
          ? data
          : {}
  }
  
  const map = {};
  const fillMap = (arr) => {
      for (let i = 0; i < arr.length; i++) {
          if (Array.isArray(arr[i])) {
            fillMap(arr[i], map)
          } else if (typeof(arr[i]) === "object") {
                const keys = Object.keys(arr[i])
                for (let j = 0; j < keys.length; j++) {
                    map[keys[j]] = arr[i][keys[j]]
                }
          }
      }
  }
  
  fillMap(data)
  return map
}
```

Reply:
@vellames-turing I could not run suggested code against the test cases successfully. It passes the tests for single key-values and fails others. It may be about not parsing the values which come as strings. How about not changing the code for now? Because it will not affect the workflow, can we think of a more elegant solution later?

Reply:
Strange, I did some tests and it was working for me, do you mind if we try to pair? So we can check it together

Reply:
Try this one please
```
    // eslint-disable-next-line camelcase
import data_stores from "../../data_stores.app.mjs";

export default {
  key: "data_stores-add-update-multiple-records",
  name: "Add or update multiple records",
  description: "Add or update multiple records to your [Pipedream Data Store](https://pipedream.com/data-stores/).",
  version: "0.0.3",
  type: "action",
  props: {
    data_stores,
    data_store: {
      propDefinition: [
        // eslint-disable-next-line camelcase
        data_stores,
        "data_store",
      ],
    },
    data: {
      label: "Data",
      type: "object",
      description: "Enter data you'd like to add as an object containing key-value pairs, or reference existing keys and update the values for those existing records. Refer to your existing keys [here](https://pipedream.com/data-stores/).",
    },
  },
  methods: {
    arrayToMap(data) {
      if (!Array.isArray(data)) {
        return typeof(data) === "object"
          ? data
          : {};
      }

      const map = {};
      const fillMap = (arr) => {
        for (let i = 0; i < arr.length; i++) {
          if (Array.isArray(arr[i])) {
            fillMap(arr[i]);
          } else if (typeof(arr[i]) === "object") {
            const keys = Object.keys(arr[i]);
            for (let j = 0; j < keys.length; j++) {
              map[keys[j]] = arr[i][keys[j]];
            }
          }
        }
      };

      fillMap(data);
      return map;
    },
  },
  async run({ $ }) {
    // Try to convert string to a valid json object
    if (typeof this.data === "string") {
      this.data = eval(`(${this.data})`);
    }
    const data = this.arrayToMap(this.data);
    const entries = Object.entries(data);
    for (const [
      key,
      value,
    ] of entries) {
      console.log(key, value);
      this.data_store.set(key, value);
    }

    if (entries.length == 1) {
      $.export("$summary", "Successfully added or updated 1 record.");
    } else {
      $.export("$summary", "Successfully added or updated " + entries.length + " records.");
    }
  },
};

```